### PR TITLE
Add support for deconstruction into field-like events.

### LINKED ISSUE #16962: Compiler crash with field-like event in deconstruction assignment
**Version Used**: 

Master branch @ February 5

**Steps to Reproduce**:

Try compiling this code:
```c#
class C {
  event System.EventHandler E;
  void M() {
    (E, _) = (null, 1);
  }
}
```

**Expected Behavior**:

Successful compilation, runtime behavior equal to `E = null;`

**Actual Behavior**:

```
Unexpected value 'EventAccess' of type 'Microsoft.CodeAnalysis.CSharp.BoundKind'
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.TransformCompoundAssignmentLHS(BoundExpression originalLHS, ArrayBuilder`1 stores, ArrayBuilder`1 temps, Boolean isDynamicAssignment) in \dotnet\Src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_CompoundAssignmentOperator.cs:line 548
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.<>c.<GetAssignmentTargetsAndSideEffects>b__176_0(BoundExpression variable, ValueTuple`4 args) in \dotnet\Src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_DeconstructionAssignmentOperator.cs:line 168
   at Microsoft.CodeAnalysis.CSharp.BoundTupleExpression.VisitAllElements[T](Action`2 action, T args) in \dotnet\Src\Compilers\CSharp\Portable\BoundTree\BoundExpression.cs:line 651
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.GetAssignmentTargetsAndSideEffects(BoundTupleExpression variables, ArrayBuilder`1 temps, ArrayBuilder`1 stores) in \dotnet\Src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_DeconstructionAssignmentOperator.cs:line 159
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitDeconstructionAssignmentOperator(BoundDeconstructionAssignmentOperator node) in \dotnet\Src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_DeconstructionAssignmentOperator.cs:line 20
   at Microsoft.CodeAnalysis.CSharp.BoundDeconstructionAssignmentOperator.Accept(BoundTreeVisitor visitor) in \dotnet\Binaries\Obj\CSharpCodeAnalysis\Debug\BoundNodes.xml.Generated.cs:line 1200
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.Visit(BoundNode node) in \dotnet\Src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 147
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node) in \dotnet\Src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 96
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(BoundExpression node) in \dotnet\Src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 225
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node) in \dotnet\Src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 206
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.VisitExpressionWithStackGuard(BoundExpression node) in \dotnet\Src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 91
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitExpressionImpl(BoundExpression node) in \dotnet\Src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter.cs:line 177
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitExpression(BoundExpression node) in \dotnet\Src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter.cs:line 151
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitUnusedExpression(BoundExpression expression) in \dotnet\Src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_ExpressionStatement.cs:line 73
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.RewriteExpressionStatement(BoundExpressionStatement node, Boolean suppressInstrumentation) in \dotnet\Src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_ExpressionStatement.cs:line 21
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitExpressionStatement(BoundExpressionStatement node) in \dotnet\Src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_ExpressionStatement.cs:line 16
   at Microsoft.CodeAnalysis.CSharp.BoundExpressionStatement.Accept(BoundTreeVisitor visitor) in \dotnet\Binaries\Obj\CSharpCodeAnalysis\Debug\BoundNodes.xml.Generated.cs:line 2708
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.Visit(BoundNode node) in \dotnet\Src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter.cs:line 140
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitBlock(BoundBlock node) in \dotnet\Src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_Block.cs:line 24
   at Microsoft.CodeAnalysis.CSharp.BoundBlock.Accept(BoundTreeVisitor visitor) in \dotnet\Binaries\Obj\CSharpCodeAnalysis\Debug\BoundNodes.xml.Generated.cs:line 2321
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.Visit(BoundNode node) in \dotnet\Src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter.cs:line 140
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.Rewrite(CSharpCompilation compilation, MethodSymbol method, Int32 methodOrdinal, NamedTypeSymbol containingType, BoundStatement statement, TypeCompilationState compilationState, SynthesizedSubmissionFields previousSubmissionFields, Boolean allowOmissionOfConditionalCalls, Boolean instrumentForDynamicAnalysis, ImmutableArray`1& dynamicAnalysisSpans, DebugDocumentProvider debugDocumentProvider, DiagnosticBag diagnostics, Boolean& sawLambdas, Boolean& sawLocalFunctions, Boolean& sawAwaitInExceptionHandler) in \dotnet\Src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter.cs:line 94
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.LowerBodyOrInitializer(MethodSymbol method, Int32 methodOrdinal, BoundStatement body, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState, Boolean instrumentForDynamicAnalysis, DebugDocumentProvider debugDocumentProvider, ImmutableArray`1& dynamicAnalysisSpans, DiagnosticBag diagnostics, VariableSlotAllocator& lazyVariableSlotAllocator, ArrayBuilder`1 lambdaDebugInfoBuilder, ArrayBuilder`1 closureDebugInfoBuilder, StateMachineTypeSymbol& stateMachineTypeOpt) in \dotnet\Src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 1170
```

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #16962.

@dotnet/roslyn-compiler, @jcouv Please review.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AssignmentOperator.cs [hunk 1]
```diff
@@ -103,6 +103,26 @@ private BoundExpression VisitAssignmentOperator(BoundAssignmentOperator node, bo
...
+                        const bool isDynamic = false;
+                        return RewriteWindowsRuntimeEventAssignmentOperator(eventAccess.Syntax,
+                                                                            eventAccess.EventSymbol,
+                                                                            EventAssignmentKind.Assignment,
+                                                                            isDynamic,
+                                                                            eventAccess.ReceiverOpt,
+                                                                            rewrittenRight);
+                    }
+
+                    // Only Windows Runtime field like events can come through here:
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AssignmentOperator.cs]: nit: field-like

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs [hunk 2]
```diff
@@ -6240,7 +6240,235 @@ public static void Main()
...
+    {
+        System.Console.WriteLine(""Handler"");
+    }
+}
+";
+
+            var comp = CompileAndVerify(source, expectedOutput:
+@"True
+Handler", additionalRefs: s_valueTupleRefs);
+            comp.VerifyDiagnostics();
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: Would it be worth checking IL for one of the scenarios?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: > Would it be worth checking IL for one of the scenarios?

I  believe behavior verification is sufficient.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AssignmentOperator.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AssignmentOperator.cs]: > nit: field-like

Thanks, I'll fix this.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need a second sign-off.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs [hunk 3]
```diff
@@ -332,15 +332,30 @@ private BoundIndexerAccess TransformIndexerAccess(BoundIndexerAccess indexerAcce
                 indexerAccess.Type);
         }
 
-        private BoundFieldAccess TransformReferenceTypeFieldAccess(BoundFieldAccess fieldAccess, BoundExpression receiver, ArrayBuilder<BoundExpression> stores, ArrayBuilder<LocalSymbol> temps)
+        /// <summary>
+        /// Returns true if the <paramref name="receiver"/> was transformed.
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs]: I think `transformed` in this spec should be `lowered and transformed`, and should give some hint as to how the method should decide whether or not it should return true or false.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs [hunk 4]
```diff
@@ -452,7 +468,10 @@ private BoundExpression TransformCompoundAssignmentLHS(BoundExpression originalL
                         var propertyAccess = (BoundPropertyAccess)originalLHS;
                         if (propertyAccess.PropertySymbol.RefKind == RefKind.None)
                         {
-                            return TransformPropertyAccess(propertyAccess, stores, temps);
+                            // CONSIDER: this is a temporary object that will be rewritten away before this lowering completes.
+                            // Mitigation: this will only produce short-lived garbage for compound assignments and increments/decrements of properties.
+                            return propertyAccess.Update(TransformPropertyOrEventReceiver(propertyAccess.PropertySymbol, propertyAccess.ReceiverOpt, stores, temps),
+                                                         propertyAccess.PropertySymbol, propertyAccess.ResultKind, propertyAccess.Type);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs]: Minor point: The corresponding comment for fields does not include `CONSIDER: ... Mitigation: ...`. Perhaps drop that here and use the simpler comment from fields here instead.

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs (hunk 4 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs]: > Perhaps drop that here and use the simpler comment from fields here instead.

Sure, I'll do this.

**Pull request merged**